### DISCUSSION TITLE
Can we retrieve the matching tokens for a document?

### DISCUSSION THREAD

**Discussion by mrisher** [xhluca/bm25s#89] [Q&A]
We're trying to display some highlighting in the UI based on which tokens match. I can't find any method that returns the matching tokens from a query, but wanted to confirm? Ideally we'd have something in the `results` object from the call `results, scores = retriever.retrieve(...)` so we would know that `document ID #123` had tokens `["fish", "cat"]`?

Thank you

**Comment by xhluca** [comment]
This is not built in, but i think it should be easy to verify? E.g.:

```python
import bm25s

query = "what are cats?"
doc = "Cats are animals"

query_tokens = set(bm25s.tokenize(query, return_ids=False)[0])
doc_tokens = bm25s.tokenize(doc, return_ids=False)[0]

matches = [1 if dt in query_tokens else 0 for dt in doc_tokens]
```

**Reply by xhluca** [reply]
Code will be more complex for other cases. however, it should be fast since oyu are only interested in the top-k instances for highlighting.

**Reply by mrisher** [reply]
Thank you. Does that also give me a way to get token positions? I would like to boost the score for exact phrase matches and don't really understand the internals of the token index.

**Reply by xhluca** [reply]
No, but I think it would be straightforward to build a scorer for reranking bm25s results. Since this would be modular and separate from bm25s, i think it would be more appropriate as a separate tool, perhaps in the `utils` module or in a separate library.

**Comment by greenersharp** [comment]
I would like to second the addition of this functionality.  

In my use case I need to make sure the documents returned by `retrieve` match some of my query tokens.
I need to make 100% sure some query tokens are matched.

I think it would be very useful to have the `retrieve` method return the matching query tokens for each result.

Thank you.

**Reply by xhluca** [reply]
Does this solution work for you?

https://github.com/xhluca/bm25s/discussions/89#discussioncomment-11388513

I think this shouldn't be in retrieve as retrieve is already complex. utils package would be better, and if it's a popular feature we could add it into retrieve in a future refactor of the retrieve function.

**Reply by xhluca** [reply]
Btw this is the retrieve function: https://github.com/xhluca/bm25s/blob/3cab2136fa27810b99bca68dd9ce0abfa8a3e72c/bm25s/__init__.py#L694-L874

almost 200 lines without counting the internal function like `_retrieve_numba_functional`. that's why i think it's cleaner to have a utils. 

feel free to contribute this feature if you believe it's important to add it.

**Reply by greenersharp** [reply]
Thanks, I looked thru the code (very nice code btw) and it seems like  `_compute_relevance_from_scores` is currently returning scores only, but it can be modified to return a tuple of `(scores,query_index_ids)`.   Then it's just a matter of modifying a few other functions to bubble the new tuple up to the `retrieve` function.

I'll work on  adding this feature and submit the PR request sometime soon.

Take care.

**Reply by xhluca** [reply]
thanks. 

i'm not sure if it makes sense to do change the signature of the outputs of internal functions, as they may be used elsewhere as well (it's been a while so don't remember on top of my head). also, you'd need to be careful with changing the signature of the output of `retrieve` as it may be a break changing. 

Finally, it will be important to think about how this change will affect the performance of the end result, as the extra operations may increase the time it takes to compute the scores, as the `_compute_relevance_from_scores_jit_ready` function is carefully written to maximize speed and seemingly simple operations may add a lot of time. To address this, i would only feel comfortable if the PR also includes the results on the [bm25-benchmarks](https://github.com/xhluca/bm25-benchmarks) compared to the latest release (perhaps on both linux and mac). If the performance end up being lower, we would need another function called `_compute_relevance_from_scores_jit_ready_with_query_index_ids` that is only used when we want the indexes, but that'd increase the complexity of the code (=harder to maintain/debug).

Overall, i'm not sure this is the best way to approach the problem, whereas an `utils` function that is applied posthoc, which only looks at the tokens that are matched, may make more sense.

I will look at the PR either way but i will consider the performance and maintanability seriously before merging, as `_compute_relevance_from_scores` and ` _compute_relevance_from_scores_jit_ready` are low-level function that are not designed to be often modified for the reason above.
